### PR TITLE
Refactor authentication into middleware

### PR DESCRIPTION
**Pull request by Xtansia**
### Description
Refactors the authentication handling of the transport out into a middleware concept. Allows pulling AWS SigV4 auth out to a separate crate.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch/src/http/middleware.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    fn pre_process_request(&self, request_builder: RequestBuilder) -> RequestBuilder {
+        request_builder
+    }
+
+    fn post_process_request(
+        &self,
+        request: Request,
+    ) -> BoxFuture<Result<Request, BoxError<'static>>> {
+        Box::pin(future::ready(Ok(request)))
+    }
```
Comment (dblock) [inline opensearch/src/http/middleware.rs:42]: The naming seems off. Is there some kind of Rust standard for middleware/callbacks? I would adopt either an event style `on_before`, `on_after` or a stack style `configure` and `call`?

**Review by dblock — COMMENTED** [review]
I think this is very good! One comment below on naming.

**Comment by Xtansia** [comment]
@dblock Yes, I'm planning on coming back to this soon-ish, especially now that the aws deps have stabilised.

**Comment by Xtansia** [comment]
Have decided to split this into multiple smaller more specific PRs

**Pull request closed without merging**